modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/factory/CETImplFactories.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.client.extension.type.internal.factory;
+
+import com.liferay.client.extension.constants.ClientExtensionEntryConstants;
+import com.liferay.client.extension.type.factory.CETImplFactory;
+import com.liferay.portal.kernel.feature.flag.FeatureFlagManagerUtil;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * @author Drew Brokke
+ */
+public class CETImplFactories {
+
+	public CETImplFactories() {
+		_cetImplFactories = HashMapBuilder.<String, CETImplFactory<?>>put(
+			ClientExtensionEntryConstants.TYPE_CUSTOM_ELEMENT,
+			new CustomElementCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_FDS_CELL_RENDERER,
+			new FDSCellRendererCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_GLOBAL_CSS,
+			new GlobalCSSCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_GLOBAL_JS,
+			new GlobalJSCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_IFRAME,
+			new IFrameCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_JS_IMPORT_MAPS_ENTRY,
+			new JSImportMapsEntryCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_STATIC_CONTENT,
+			new StaticContentCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_THEME_CSS,
+			new ThemeCSSCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_THEME_SPRITEMAP,
+			new ThemeSpritemapCETImplFactoryImpl()
+		).put(
+			ClientExtensionEntryConstants.TYPE_THEME_FAVICON,
+			new ThemeFaviconCETImplFactoryImpl()
+
+			// TODO
+
+		/*).put(
+			ClientExtensionEntryConstants.TYPE_THEME_JS,
+			new ThemeJSCETImplFactoryImpl()*/
+
+		).build();
+	}
+
+	public CETImplFactory<?> getCETImplFactory(String key) {
+		if (_isDisabled(key)) {
+			return null;
+		}
+
+		return _cetImplFactories.get(key);
+	}
+
+	public Set<String> getTypes() {
+		SortedSet<String> types = new TreeSet<>(_cetImplFactories.keySet());
```

Comment:
Why sorted? Isn't it a waste of CPU time adn RAM? :thinking:

Reply:
Also, in the previous code `_type` was set once and reused all the time which made it unnecessary to create new objects every time `getTypes()` is invoked.

Reply:
Oh yes, you are right. It was sorted [here](https://github.com/drewbrokke/liferay-portal/pull/1445/files#diff-01abd0524cce98e01755d244f6e13bb7b8faa0d7d4506ee1c6ce4ff2b787d0e6L95)...

So, this code is even worse than I thought :sweat_smile: .

Where we had a [constant time operation with no RAM comsumption](https://github.com/drewbrokke/liferay-portal/pull/1445/files#diff-01abd0524cce98e01755d244f6e13bb7b8faa0d7d4506ee1c6ce4ff2b787d0e6L153) now we have to:

    1. Clone a set just to return it (more RAM)
    2. Sort the cloned set
    3. Filter it
    
Also, remember that the filtering is invoking `isDisabled` which has the chained ifs. Thus, if you had 500 CX types, you need to execute 500 * 250 ifs on average :grimacing: